### DISCUSSION TITLE
Move to gcc 10 for Node.js 20?

### DISCUSSION THREAD

**Discussion by richardlau** [nodejs/help#5015] [Ideas]
Node.js 20 is [scheduled for April 2023](https://github.com/nodejs/Release).

We are currently building most of the Unix-like platforms with gcc 8 (macOS and FreeBSD use clang). Furthermore we've started to see issues with bugs/unsupported features with gcc 8 in upstream V8:
- https://chromium-review.googlesource.com/c/v8/v8/+/3988005
- https://chromium-review.googlesource.com/c/v8/v8/+/4097624

which are unlikely to be fixed as [gcc 8 is no longer supported by gnu](https://gcc.gnu.org/gcc-8/).

In Node.js itself we've seen an issue on arn64 where gcc deprecated one compile option (`-msign-return-address=all`) for another (`-mbranch-protection`, introduced in gcc 9): https://github.com/nodejs/node/pull/45756#issuecomment-1339620333

Furthermore, upstream V8 are also moving to C++20 (we have https://github.com/nodejs/node/issues/45402 open to discuss the same for Node.js. [C++20 feature availability](https://gcc.gnu.org/projects/cxx-status.html#cxx20) is better with more recent versions of gcc.

Semver major boundaries are the natural place to bump build requirements. Thoughts on bumping the minimum supported gcc compiler for Node.js 20 to gcc 10?

I have checked and gcc 10 is available on the platforms that we are currently building with gcc 8 (some platforms do not have gcc 11).

In terms of practicalities this would involve:
- Updating documentation in https://github.com/nodejs/node/blob/main/BUILDING.md.
- Updating the warning in https://github.com/nodejs/node/blob/464e4d293aafc61b69a047363e057473ad998f20/configure.py#L1023-L1025. This is just an advisory warning -- we do not actively prevent anyone from trying to compile with older gcc compilers.
- Corresponding work for @nodejs/build to update the CI machines where necessary.
  - For the RHEL/Fedora based systems this will involve installing gcc-toolset-10.
  - For Debian/Ubuntu based systems this will involve installing gcc-10.

**Comment by mhdawson** [comment]
Given that we have gcc 10 available for all of our platforms using gcc, it sounds like it is reasonable to plan to move to gcc 10 for Node.js 20.

**Comment by KidusB9** [comment]
i think It seems that upgrading the minimum supported version of gcc to 10 is necessary for Node.js 20 due to issues with unsupported features and bugs in the older gcc 8 compiler, as well as the desire to take advantage of the improved C++20 feature availability in more recent versions of gcc. This upgrade would involve updating the documentation and warning in the build process, as well as updating the CI machines used for testing. It may also require installing newer versions of gcc on certain systems.
